### DISCUSSION TITLE
Suppressing CS0618 obsolete when referencing from a obsolete member's attribute?

### DISCUSSION THREAD

**Discussion by acaly** [dotnet/csharplang#4391] [Q&A]
Let's say I have a Entity Framework model like this:

```c#
    [ForeignKey(nameof(User))]
    public int UserID { get; set; }
    public User User { get; set; }
```

Now I would like to change to use another implementation of user so I marked the `User` class, `User` property, and `UserID` property in the above class as `[Obsolete]`. The reference of the `User` class in the `User` property no longer shows the CS0618 warning because the property itself is also obsolete, but the `nameof(User)` still shows the warning, although the `UserID` is also obsolete. I would expect this warning to go off.

Is there any reason it was not designed as so?

**Comment by huoyaoyuan** [comment]
Another thing to consider is `ObsoleteAttribute` with alternate id and url.

**Comment by RikkiGibson** [comment]
This just looks to me like use of an obsolete symbol within an obsolete context. That shouldn't produce obsolete diagnostics.

The specification doesn't actually mention that obsolete diagnostics are suppressed when used inside an obsolete context. Therefore it seems like an implementation defined behavior and that the scenario you've outlined is a compiler bug.
https://github.com/dotnet/csharplang/blob/master/spec/attributes.md#the-obsolete-attribute

**Reply by RikkiGibson** [reply]
I have filed dotnet/roslyn#51071 to address this behavior.

**Reply by acaly** [reply]
Thanks!

**Comment by acaly** [comment]
Since the issue dotnet/roslyn#51071 has been closed, I really want to continue the discussion here. (According to the explanation in that issue, csharplang should be a better place to discuss this instead of roslyn.)

If anyone thinks it's necessary I can write a proposal for this. I think the objectives and requirements are quite clear.
